// @vaadin/hilla-file-router/runtime/createMenuItems.js
function __REGISTER__(feature, vaadinObj = window.Vaadin ??= {}) {
  vaadinObj.registrations ??= [];
  vaadinObj.registrations.push({
    is: feature ? `${"@vaadin/hilla-file-router"}/${feature}` : "@vaadin/hilla-file-router",
    version: "24.5.5"
  });
}
import { signal } from "@vaadin/hilla-react-signals";
const viewsSignal = signal(window.Vaadin?.views);
function createMenuItems() {
  __REGISTER__("createMenuItems", window.Vaadin);
  const collator = new Intl.Collator("en-US");
  if (!viewsSignal.value) {
    return [];
  }
  return Object.entries(viewsSignal.value).filter(([_, value]) => !value.menu?.exclude).map(([path, config]) => ({
    to: path,
    icon: config.menu?.icon,
    title: config.menu?.title ?? config.title,
    order: config.menu?.order
  })).sort((menuA, menuB) => {
    const ordersDiff = (menuA.order ?? Number.MAX_VALUE) - (menuB.order ?? Number.MAX_VALUE);
    return ordersDiff !== 0 ? ordersDiff : collator.compare(menuA.to, menuB.to);
  });
}
if (import.meta.hot) {
  import.meta.hot.on("fs-route-update", () => {
    fetch("?v-r=routeinfo").then(async (resp) => resp.json()).then((json) => {
      viewsSignal.value = json;
    }).catch((e) => {
      console.error("Failed to fetch route info", e);
    });
  });
}
export {
  createMenuItems,
  viewsSignal
};
